File: Java/Mng.java
package mng;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class Mng {
    private static final byte[] MNG_SIGNATURE = {(byte) 0x8a, 0x4d, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
    private static final byte[] PNG_SIGNATURE = {(byte) 0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};

    public static final class Frame {
        public final byte[] png;
        public final long delayMilliseconds;

        public Frame(byte[] png, long delayMilliseconds) {
            this.png = png;
            this.delayMilliseconds = delayMilliseconds;
        }
    }

    public static final class Animation {
        public final long width;
        public final long height;
        public final long ticksPerSecond;
        public final long nominalLayerCount;
        public final long nominalFrameCount;
        public final long nominalPlayTime;
        public final long simplicityProfile;
        public final List<Frame> frames;

        private Animation(long width, long height, long ticksPerSecond, long nominalLayerCount, long nominalFrameCount, long nominalPlayTime, long simplicityProfile, List<Frame> frames) {
            this.width = width;
            this.height = height;
            this.ticksPerSecond = ticksPerSecond;
            this.nominalLayerCount = nominalLayerCount;
            this.nominalFrameCount = nominalFrameCount;
            this.nominalPlayTime = nominalPlayTime;
            this.simplicityProfile = simplicityProfile;
            this.frames = Collections.unmodifiableList(frames);
        }

        public int frameIndexAt(Duration elapsed) {
            if (frames.isEmpty()) return -1;
            if (frames.size() == 1) return 0;
            long duration = 0;
            for (Frame frame : frames) duration += frame.delayMilliseconds;
            if (duration <= 0) return 0;
            long milliseconds = elapsed.toMillis() % duration;
            if (milliseconds < 0) milliseconds += duration;
            for (int index = 0; index < frames.size(); index++) {
                if (milliseconds < frames.get(index).delayMilliseconds) return index;
                milliseconds -= frames.get(index).delayMilliseconds;
            }
            return frames.size() - 1;
        }
    }

    public static final class ParseException extends Exception {
        public ParseException(String message) {
            super(message);
        }
    }

    private Mng() {
    }

    public static boolean isMng(byte[] data) {
        if (data == null || data.length < MNG_SIGNATURE.length) return false;
        for (int index = 0; index < MNG_SIGNATURE.length; index++) if (data[index] != MNG_SIGNATURE[index]) return false;
        return true;
    }

    public static Animation parse(byte[] data) throws ParseException {
        if (!isMng(data)) throw new ParseException("not an MNG image");
        long width = 0;
        long height = 0;
        long ticksPerSecond = 100;
        long nominalLayerCount = 0;
        long nominalFrameCount = 0;
        long nominalPlayTime = 0;
        long simplicityProfile = 0;
        List<Frame> frames = new ArrayList<>();
        ByteArray frame = new ByteArray();
        Long pendingDelay = null;
        boolean inPng = false;
        int offset = MNG_SIGNATURE.length;
        while (offset < data.length) {
            if (data.length - offset < 12) throw new ParseException("truncated MNG chunk");
            long dataLength = readU32(data, offset);
            long chunkLength = dataLength + 12;
            if (chunkLength < 12 || chunkLength > data.length - offset) throw new ParseException("invalid MNG chunk length");
            String type = new String(data, offset + 4, 4, java.nio.charset.StandardCharsets.US_ASCII);
            if (type.equals("MHDR")) {
                if (dataLength < 28) throw new ParseException("invalid MNG header");
                width = readU32(data, offset + 8);
                height = readU32(data, offset + 12);
                long ticks = readU32(data, offset + 16);
                if (ticks > 0) ticksPerSecond = ticks;
                nominalLayerCount = readU32(data, offset + 20);
                nominalFrameCount = readU32(data, offset + 24);
                nominalPlayTime = readU32(data, offset + 28);
                simplicityProfile = readU32(data, offset + 32);
            } else if (type.equals("FRAM")) {
                pendingDelay = parseFrameDelay(data, offset + 8, (int) dataLength);
            }
            if (!inPng && type.equals("IHDR")) {
                frame.write(PNG_SIGNATURE);
                inPng = true;
            }
            if (inPng) {
                frame.write(data, offset, (int) chunkLength);
                if (type.equals("IEND")) {
                    long delay = pendingDelay == null ? 100 : Math.max(0, Math.round((double) pendingDelay * 1000.0 / ticksPerSecond));
                    frames.add(new Frame(frame.toByteArray(), delay));
                    frame.clear();
                    pendingDelay = null;
                    inPng = false;
                }
            }
            offset += (int) chunkLength;
        }
        if (inPng || frames.isEmpty()) throw new ParseException("MNG contains no complete frames");
        return new Animation(width, height, ticksPerSecond, nominalLayerCount, nominalFrameCount, nominalPlayTime, simplicityProfile, frames);
    }

    public static byte[] firstPngFrame(byte[] data) throws ParseException {
        return parse(data).frames.get(0).png.clone();
    }

    private static long readU32(byte[] data, int offset) {
        return ((long) (data[offset] & 0xff) << 24) | ((long) (data[offset + 1] & 0xff) << 16) | ((long) (data[offset + 2] & 0xff) << 8) | (data[offset + 3] & 0xffL);
    }

    private static Long parseFrameDelay(byte[] data, int offset, int length) {
        if (length == 0) return null;
        int index = 1;
        while (index < length && data[offset + index] != 0) index++;
        if (index == length) return null;
        index++;
        if (index >= length) return null;
        boolean hasDelay = data[offset + index++] != 0;
        if (length - index < 3) return null;
        index += 3;
        if (!hasDelay || length - index < 4) return null;
        return readU32(data, offset + index);
    }

    private static final class ByteArray {
        private byte[] data = new byte[64];
        private int size;

        void write(byte[] source) {
            write(source, 0, source.length);
        }

        void write(byte[] source, int offset, int length) {
            ensure(size + length);
            System.arraycopy(source, offset, data, size, length);
            size += length;
        }

        byte[] toByteArray() {
            byte[] result = new byte[size];
            System.arraycopy(data, 0, result, 0, size);
            return result;
        }

        void clear() {
            size = 0;
        }

        private void ensure(int required) {
            if (required <= data.length) return;
            int capacity = data.length;
            while (capacity < required) capacity *= 2;
            byte[] resized = new byte[capacity];
            System.arraycopy(data, 0, resized, 0, size);
            data = resized;
        }
    }
}

File: Kotlin/Mng.kt
package mng

import java.io.ByteArrayOutputStream
import java.time.Duration

data class MngFrame(val png: ByteArray, val delayMilliseconds: Long)

class MngAnimation(
    val width: Long,
    val height: Long,
    val ticksPerSecond: Long,
    val nominalLayerCount: Long,
    val nominalFrameCount: Long,
    val nominalPlayTime: Long,
    val simplicityProfile: Long,
    val frames: List<MngFrame>
) {
    fun frameIndexAt(elapsed: Duration): Int {
        if (frames.isEmpty()) return -1
        if (frames.size == 1) return 0
        val duration = frames.sumOf { it.delayMilliseconds }
        if (duration <= 0) return 0
        var milliseconds = elapsed.toMillis() % duration
        if (milliseconds < 0) milliseconds += duration
        for ((index, frame) in frames.withIndex()) {
            if (milliseconds < frame.delayMilliseconds) return index
            milliseconds -= frame.delayMilliseconds
        }
        return frames.lastIndex
    }
}

class MngException(message: String) : Exception(message)

object Mng {
    private val mngSignature = byteArrayOf(0x8a.toByte(), 0x4d, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a)
    private val pngSignature = byteArrayOf(0x89.toByte(), 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a)

    fun isMng(data: ByteArray): Boolean = data.size >= mngSignature.size && data.copyOfRange(0, mngSignature.size).contentEquals(mngSignature)

    @Throws(MngException::class)
    fun parse(data: ByteArray): MngAnimation {
        if (!isMng(data)) throw MngException("not an MNG image")
        var width = 0L
        var height = 0L
        var ticksPerSecond = 100L
        var nominalLayerCount = 0L
        var nominalFrameCount = 0L
        var nominalPlayTime = 0L
        var simplicityProfile = 0L
        val frames = mutableListOf<MngFrame>()
        val frame = ByteArrayOutputStream()
        var pendingDelay: Long? = null
        var inPng = false
        var offset = mngSignature.size
        while (offset < data.size) {
            if (data.size - offset < 12) throw MngException("truncated MNG chunk")
            val dataLength = data.readU32(offset).toInt()
            val chunkLength = dataLength + 12
            if (chunkLength < 12 || chunkLength > data.size - offset) throw MngException("invalid MNG chunk length")
            val type = data.copyOfRange(offset + 4, offset + 8).toString(Charsets.US_ASCII)
            val chunkData = data.copyOfRange(offset + 8, offset + 8 + dataLength)
            if (type == "MHDR") {
                if (dataLength < 28) throw MngException("invalid MNG header")
                width = data.readU32(offset + 8)
                height = data.readU32(offset + 12)
                val ticks = data.readU32(offset + 16)
                if (ticks > 0) ticksPerSecond = ticks
                nominalLayerCount = data.readU32(offset + 20)
                nominalFrameCount = data.readU32(offset + 24)
                nominalPlayTime = data.readU32(offset + 28)
                simplicityProfile = data.readU32(offset + 32)
            } else if (type == "FRAM") {
                pendingDelay = parseFrameDelay(chunkData)
            }
            if (!inPng && type == "IHDR") {
                frame.write(pngSignature)
                inPng = true
            }
            if (inPng) {
                frame.write(data, offset, chunkLength)
                if (type == "IEND") {
                    val delay = pendingDelay?.let { Math.max(0L, Math.round(it * 1000.0 / ticksPerSecond)) } ?: 100L
                    frames += MngFrame(frame.toByteArray(), delay)
                    frame.reset()
                    pendingDelay = null
                    inPng = false
                }
            }
            offset += chunkLength
        }
        if (inPng || frames.isEmpty()) throw MngException("MNG contains no complete frames")
        return MngAnimation(width, height, ticksPerSecond, nominalLayerCount, nominalFrameCount, nominalPlayTime, simplicityProfile, frames)
    }

    fun firstPngFrame(data: ByteArray): ByteArray = parse(data).frames.first().png.copyOf()

    private fun parseFrameDelay(data: ByteArray): Long? {
        if (data.isEmpty()) return null
        var offset = 1
        while (offset < data.size && data[offset].toInt() != 0) offset++
        if (offset == data.size) return null
        offset++
        if (offset >= data.size) return null
        val hasDelay = data[offset].toInt() != 0
        offset++
        if (data.size - offset < 3) return null
        offset += 3
        if (!hasDelay || data.size - offset < 4) return null
        return data.readU32(offset)
    }

    private fun ByteArray.readU32(offset: Int): Long = ((this[offset].toLong() and 0xff) shl 24) or ((this[offset + 1].toLong() and 0xff) shl 16) or ((this[offset + 2].toLong() and 0xff) shl 8) or (this[offset + 3].toLong() and 0xff)
}
